src/buffer_manager.ts

```diff
@@ -612,18 +675,21 @@ export class BufferManager implements Disposable {
             vim.api.nvim_buf_set_var(bufId, "vscode_editor_options", vscode_editor_options)
             vim.api.nvim_buf_set_var(bufId, "vscode_uri", docUri)
             vim.api.nvim_buf_set_var(bufId, "vscode_uri_data", docUriJson)
-            vim.api.nvim_buf_set_name(bufId, bufname)
             vim.api.nvim_buf_set_option(bufId, "modifiable", not isExternalDoc)
-            -- force nofile, just in case if the buffer was created externally
-            vim.api.nvim_buf_set_option(bufId, "buftype", "nofile")
+            -- force acwrite, which is similar to nofile, but will only be written via the BufWriteCmd autocommand.
+            vim.api.nvim_buf_set_option(bufId, "buftype", "acwrite")
             vim.api.nvim_buf_set_option(bufId, "buflisted", true)
+
+            if bufname ~= vim.NIL then
+                vim.api.nvim_buf_set_name(bufId, bufname)
+            end
         `,
             [
                 bufId,
                 lines,
                 makeEditorOptionsVariable(editor?.options),
-                docUri.toString(),
-                docUri.toJSON(),
+                document.uri.toString(),
+                document.uri.toJSON(),
                 bufname,
                 this.isExternalTextDocument(document),
             ],
```

Comment:
Perhaps we can take this opportunity to encapsulate these contents into 'lua/vscode/internal.lua'.